pkg/table/tables/tables.go

```diff
@@ -576,21 +576,9 @@ func (t *TableCommon) UpdateRecord(ctx context.Context, sctx table.MutateContext
 	}
 	sessVars := sctx.GetSessionVars()
 	// rebuild index
-	if !sessVars.InTxn() {
-		savePresumeKeyNotExist := sessVars.PresumeKeyNotExists
-		if !sessVars.ConstraintCheckInPlace && sessVars.TxnCtx.IsPessimistic {
-			sessVars.PresumeKeyNotExists = true
-		}
-		err = t.rebuildIndices(sctx, txn, h, touched, oldData, newData, table.WithCtx(ctx))
-		sessVars.PresumeKeyNotExists = savePresumeKeyNotExist
-		if err != nil {
-			return err
-		}
-	} else {
-		err = t.rebuildIndices(sctx, txn, h, touched, oldData, newData, table.WithCtx(ctx))
-		if err != nil {
-			return err
-		}
+	err = t.rebuildIndices(sctx, txn, h, touched, oldData, newData, table.WithCtx(ctx))
```

Comment:
Why change these lines? The test still passes if these lines are unchanged.

Reply:
Because this code is unnecessary now. The current `LazyCheckKeyNotExists` is:

```
func (s *SessionVars) LazyCheckKeyNotExists() bool {
	if s.StmtCtx.ErrGroupLevel(errctx.ErrGroupDupKey) != errctx.LevelError {
		// This branch means we are in `insert/update ignore`.
		// The executor will handle the dup-key error and ignore it in executor,
		// so we must check the dup-key error in place to make sure the executor can get the error.
		return false
	}
	return s.PresumeKeyNotExists || (s.TxnCtx != nil && s.TxnCtx.IsPessimistic)
}
```

You can see that if `sessVars.TxnCtx.IsPessimistic` is true, `s.PresumeKeyNotExists || (s.TxnCtx != nil && s.TxnCtx.IsPessimistic)` will always return true, we do not need to set `PresumeKeyNotExists`.

Reply:
reverted diffs in `LazyCheckKeyNotExists` can also fix this bug...

Reply:
‌‌Let's go back to the original place. The purpose of `tidb_constraint_check_in_place` is only to make the lazy unique index check for insert effective. Historically, only inserts had a lazy unique index check. `tidb_constraint_check_in_place` was actually designed to remove this exclusive check for inserts, so that all statements execute consistent in-place checks. Therefore, update statements have never had a lazy check. So even if `tidb_constraint_check_in_place` is off, updates should still report an error. @lcwangchao @ekexium

Reply:
As for why the name was designed this way, that was the product manager's decision at the time. However, I don't think it's necessary to make updates lazy check as well. The reason insert needs lazy check is because it performs more efficiently during bulk load. Bulk loading should rarely involve updates, right?

Reply:
If you want the update to also perform a lazy check, handling it only at https://github.com/pingcap/tidb/blob/08147e7e2586b1db7162ad9c2dfc134b5a5f6b43/pkg/table/tables/tables.go#L579-L589 is not enough. It also requires the 2PC phase to first check for unique constraints like insert before committing. This way, for https://github.com/pingcap/tidb/issues/54489, there will be no retry during an update; instead, a duplicate entry error reported by TiKV will be received.